aten/src/ATen/native/cpu/zmath.h

```diff
@@ -138,6 +138,16 @@ inline c10::complex<double> ceil_impl (c10::complex<double> z) {
   return c10::complex<double>(std::ceil(z.real()), std::ceil(z.imag()));
 }
 
+inline c10::complex<float> sgn_impl (c10::complex<float> z) {
+  auto angle = std::arg(z);
+  return c10::complex<float>(std::cos(angle), std::sin(angle));
```

Comment:
I actually find this is a pretty unexpected way of computing the normalized vector. Ordinarily, I would have expected you compute the absolute value and then divide by that. Trigonometric functions are generally more expensive, and here you're computing THREE trig functions (atan2, cos, sin), whereas computing the norm and then dividing would just require a square, addition, square root and then division.

Reply:
I agree but this function is is a more numerically robust version of x/abs(x) (which is why we decided to add this function in the first place)

Reply:
rsqrt, no division (if you don't care about precision)!